storage/SMAIndexSubBlock.hpp

```diff
@@ -0,0 +1,432 @@
+
+  /**
+   * @brief Updates the index to reflect the removal of a single tuple.
+   *
+   * @param tuple The id of the tuple which is going to be removed (i.e. it still
+   *        exists within the storage block.
+   */
+  void removeEntry(const tuple_id tuple) override;
+
+  /**
+   * @brief Updates the index to reflect the addition of several tuples.
+   *
+   * @param tuples The ids of the tuple which have been added (i.e. they exist within
+   *        the storage block).
+   *
+   * @return \c true if successful.
+   */
+  bool bulkAddEntries(const TupleIdSequence &tuples) override;
+
+  /**
+   * @brief Updates the index to reflect the removal of several tuples.
+   *
+   * @param tuples The ids of the tuple which is going to be removed (i.e. they
+   *        still exist within the storage block.
+   */
+  void bulkRemoveEntries(const TupleIdSequence &tuples) override;
+
+  /**
+   * @brief Gives an estimate of how long it will take to respond to a query.
+   * 
+   * The SMA index will detect if one of the following cases is true:
+   *   1) Complete match: all the tuples in this subblock will match the predicate
+   *   2) Empty match:    none of the tuples will match
+   *   3) Partial match:  some of the tuples may match
+   * If there is a partial match, the SMA index is of no use. However, in a
+   * Complete or Empty match, the SMA index can speed up the selection process
+   * and should be used.
+   *
+   * @param predicate A simple predicate too be evaluated.
+   * 
+   * @return The cost associated with the type of match. Empty matches are constant,
+   *         partial matches require a scan, and complete matches require something
+   *         less than a regular scan (because we don't need to do the comparison
+   *         operation for each tuple) but is still linear with the number of tuples.
+   */
+  predicate_cost_t estimatePredicateEvaluationCost(
+      const ComparisonPredicate &predicate) const override;
+
+  /**
+   * @warning Calling this method on the SMA index implies that we are not going
+   *          to do a scan for some tuple matches. As in, the SMA index will
+   *          either return an empty set of tuple ids, or a set of tuple ids
+   *          which is the entire set of all tuple ids in the storage subblock.
+   * @note Currently this version only supports simple comparisons of a literal
+   *       value with a non-composite key.
+   **/
+  TupleIdSequence* getMatchesForPredicate(const ComparisonPredicate &predicate,
+                                          const TupleIdSequence *filter) const override;
+
+  /**
+   * @brief Update the index to reflect the current state of the storage block.
+   *
+   * @return \c true if successful.
+   */
+  bool rebuild() override;
+
+  /**
+   * @brief Returns if the index is consistent. Rebuilding will ensure this
+   *        returns true.
+   *
+   * @return \c true if inconsistent (rebuild to return true).
+   */
+  bool requiresRebuild() const;
```

Comment:
Addressed. External callers are the unit tests but they are friends of the SMA so being public is unnecessary.